### PR TITLE
Mandatory on Manufacturing 

### PR DESCRIPTION
**Pull request by metas-rc**
https://github.com/metasfresh/metasfresh/issues/13571

### CODE REVIEW

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/pporder/api/HUPPOrderIssueReceiptCandidatesProcessor.java [hunk 1]
```diff
@@ -127,6 +128,7 @@ public static HUPPOrderIssueReceiptCandidatesProcessor newInstance()
 	private final transient IHUPPCostCollectorBL huPPCostCollectorBL = Services.get(IHUPPCostCollectorBL.class);
 	private final transient IHUPPOrderQtyDAO huPPOrderQtyDAO = Services.get(IHUPPOrderQtyDAO.class);
 	private final transient IWarehouseDAO warehousesRepo = Services.get(IWarehouseDAO.class);
+	private final IHUAttributesBL huAttributesBL = Services.get(IHUAttributesBL.class);
```
Comment (metas-ts) [inline backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/pporder/api/HUPPOrderIssueReceiptCandidatesProcessor.java:131]:
```suggestion
	private final transient IHUAttributesBL huAttributesBL = Services.get(IHUAttributesBL.class);
```

**Review by metas-ts — APPROVED** [review]

File: backend/de.metas.business/src/main/java/org/adempiere/mm/attributes/api/impl/AttributesBL.java [hunk 2]
```diff
@@ -212,13 +214,20 @@ public boolean hasAttributeAssigned(@NonNull final ProductId productId, @NonNull
...
+	public boolean isMandatoryOn(@NonNull final ProductId productId,
+			@NonNull final AttributeId attributeId,
+			@NonNull AttributeSourceDocument attributeSourceDocument)
 	{
 		final AttributeSetId attributeSetId = productsService.getAttributeSetId(productId);
+		final Function<AttributeSetAttribute, OptionalBoolean> mandatoryOnFunction = attributeSourceDocument.isMaterialReceipt() ?
+				AttributeSetAttribute::getMandatoryOnReceipt :
+				AttributeSetAttribute::getMandatoryOnManufacturing;
+
 		final Boolean mandatoryOnReceipt = attributesRepo.getAttributeSetAttributeId(attributeSetId, attributeId)
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/org/adempiere/mm/attributes/api/impl/AttributesBL.java:226]: this is pretty weird. to make it a bit clearer:
* inline the mandatoryOnFunction
*rename mandatoryOnRecpit to simply `mandatory`

.... but the best would be to have something like:
```java
final AttributeSetAttribute attribute = attributesRepo.getAttributeSetAttributeId(attributeSetId, attributed);
final Boolean mandatory;
if (attributeSourceDocument.isReciept()) {
mandatory = attribute.getMandatoryOnReciept().orElse(null);
}
else if (attributeSourceDocument.isManufacturing()) {
mandatory = attribute.getMandatoryOnManufacturing().orElse(null);
}
else if (attributeSourceDocument.isPicking()) {
mandatory = attribute.getMandatoryOnPicking().orElse(null);
}
else {
throw new AdempiereException("Unknown: "+attributeSourceDocument);
}
```

File: backend/de.metas.business/src/main/java/org/adempiere/mm/attributes/api/AttributeSourceDocument.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ */
+
+package org.adempiere.mm.attributes.api;
+
+public enum AttributeSourceDocument
+{
+	MaterialReceipt("R"),
+	ManufacturingOrder("M");
+
+	private final String type;
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/org/adempiere/mm/attributes/api/AttributeSourceDocument.java:30]: why do we need the "String type"?

File: backend/de.metas.business/src/main/java/org/adempiere/mm/attributes/api/impl/AttributesBL.java [hunk 4]
```diff
@@ -256,6 +265,19 @@ public ImmutableList<I_M_Attribute> getAttributesMandatoryOnPicking(final Produc
 		return attributesMandatoryOnPicking;
 	}
 
+	@Override
+	public ImmutableList<I_M_Attribute> getAttributesMandatoryOnManufacturing(final ProductId productId)
+	{
+		final AttributeSetId attributeSetId = productBL.getAttributeSetId(productId);
+		final ImmutableList<I_M_Attribute> attributesMandatoryOnManufacturing = attributesRepo.getAttributesByAttributeSetId(attributeSetId).stream()
+				.filter(attribute -> isMandatoryOn(productId,
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/org/adempiere/mm/attributes/api/impl/AttributesBL.java:273]: this feels like SQL N+1 issue... but i see other methods are doing the same.... so leave it like it is for now.

**Pull request merged**